File: sol/solution/0101-0200/0103/Solution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun zigzagLevelOrder(root: TreeNode?): List<List<Int>> {
        val levelOrder = mutableListOf<MutableList<Int>>()
        getLevelOrder(root, levelOrder, 0)
        return levelOrder
    }
    fun getLevelOrder(root: TreeNode?, levelOrder: MutableList<MutableList<Int>>, height: Int) {
        if (root == null) return
        if (height == levelOrder.size) 
            levelOrder.add(mutableListOf<Int>())
        if (height % 2 == 0) 
            levelOrder[height].add(root.`val`)  
        else 
            levelOrder[height].add(0, root.`val`)        
        getLevelOrder(root.left, levelOrder, height + 1)
        getLevelOrder(root.right, levelOrder, height + 1)
    }
}

File: sol/solution/0101-0200/0103/Solution.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
    def zigzagLevelOrder(root: TreeNode): List[List[Int]] = {
        if (root == null) return List()
        import scala.collection.mutable.Queue
        val result = scala.collection.mutable.ListBuffer[List[Int]]()
        val queue = Queue[TreeNode]()
        queue.enqueue(root)
        var leftToRight = true
        while (queue.nonEmpty) {
            val size = queue.size
            val level = new Array[Int](size)
            for (i <- 0 until size) {
                val node = queue.dequeue()
                val idx = if (leftToRight) i else size - 1 - i
                level(idx) = node.value
                if (node.left != null) queue.enqueue(node.left)
                if (node.right != null) queue.enqueue(node.right)
            }
            result += level.toList
            leftToRight = !leftToRight
        }
        result.toList
    }
}